### PR TITLE
[UC21#93#95#96#97#179] Exchange Place creation and Implementation of RPC to fetch proposals from db

### LINKED ISSUE #93: Creare ExchangesPlace e pensare all'URI per vedere la proposta di scambio (URI: #Exchanges/{id}) [UC21]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [ ]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### LINKED ISSUE #95: Creare RPC per popolare la pagina della proposta di scambio [UC21]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [ ]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### LINKED ISSUE #96: Creare ExchangeActivity e scegliere se utilizzare NewExchangeView o creare una nuova view per visualizzare proposta di scambio [UC21]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [ ]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### LINKED ISSUE #97: Popolare ExchangeActivity e la view ad essa associata con i dati ottenuti dalla RPC getExchangeById [UC21]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [x]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### LINKED ISSUE #179: Non rendere cliccabili le carte relative alla proposta di scambio [UC21]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [x]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### PR DESCRIPTION
**Pull request by Matte0Sacc0**
This PR implements all the tasks relative to UC21 #93 #95 #96 #97 #179 

Has been created a new Object ProposalPayload to allow the RCP developed for issue #95 to communicate the data to the client in one shot, avoiding multiple fetches to retrive two lists by the same proposal on the db.

In fact the issue #97 was requested to fetch the data from the db giving as a parameter the proposal ID, but had to retrive three different information:
- The receiver email
- The List of PhysicalCardsWithName from the List of PhysicalCard of the receiver user of the proposal
- The List of PhysicalCardsWithName from the List of PhysicalCard of the sender user of the proposal

By using the ProposalPayload object this can be done with a single call to db.

Moreover the issues #93 and #96 had been completed in a single istance because I've decided to reuse NewExchangeView and only change the associated Presenter and Place

In conclusion of this session, has been added the issue #179 because with the team we have together noticed the problem it wants to fix: the cards in the deck displayed to the user in the Proposal acceptance page were clickable and they don't have to.

### CODE REVIEW

**Review by dstops — APPROVED** [review]

File: src/main/java/com/aadm/cardexchange/client/views/NewExchangeViewImpl.java [hunk 1]
```diff
@@ -1,54 +1,101 @@
...
 
-    public void setReceiverDeckName(String receiverUserEmail) {
-        receiverDeck.setDeckName(receiverUserEmail);
+    @Override
+    public void setData(boolean clickable, String title, String subtitle) {
+        pageTitle.setInnerText(title);
+        pageSubtitle.setInnerText(subtitle);
+        exchangeDecks.clear();
+        senderDeck = new DeckWidget(null, null, clickable ? this : null, null, "Your Owned Cards");
+        receiverDeck = new DeckWidget(null, null, clickable ? this : null, null, "");
```
Comment (alessioarcara) [inline src/main/java/com/aadm/cardexchange/client/views/NewExchangeViewImpl.java]: That could be replaced with factory method

File: src/main/java/com/aadm/cardexchange/client/places/ExchangePlace.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    private final int proposalId;
+
+    public ExchangePlace(int proposalId) {
+        this.proposalId = proposalId;
+    }
+
+    public int getExchangeProposalId() {
+        return proposalId;
+    }
+}
```
Comment (alessioarcara) [inline src/main/java/com/aadm/cardexchange/client/places/ExchangePlace.java]: As I told you earlier, using the concepts of REST APIs, this should not be a new link. Instead, it could be expressed using "exchanges/{idexchange}" avoiding the need to create a new endpoint and a new route constant, and managing everything within the exchanges case

File: src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java [hunk 3]
```diff
@@ -7,22 +7,23 @@
...
+import java.util.*;
 
 
 public class ExchangeServiceImpl extends RemoteServiceServlet implements ExchangeService, MapDBConstants {
     private static final long serialVersionUID = 5868088467963819042L;
     private final MapDB db;
     private final Gson gson = new Gson();
 
+    private final Type proposalType = new TypeToken<Proposal>() {}.getType();
+
```
Comment (alessioarcara) [inline src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java]: This type is unnecessary and incorrect

File: src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java [hunk 4]
```diff
@@ -53,8 +54,59 @@ public boolean addProposal(String token, String receiverUserEmail, List<Physical
             throw new InputException("Invalid receiver physical cards");
         }
         Proposal newProposal = new Proposal(email, receiverUserEmail, senderPhysicalCards, receiverPhysicalCards);
-        Map<Integer, Proposal> proposalMap = db.getPersistentMap(getServletContext(), PROPOSAL_MAP_NAME, Serializer.INTEGER, new GsonSerializer<>(gson));
+        Map<Integer, Proposal> proposalMap = db.getPersistentMap(getServletContext(), PROPOSAL_MAP_NAME, Serializer.INTEGER, new GsonSerializer<>(gson, proposalType));
```
Comment (alessioarcara) [inline src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java]: This type is unnecessary and incorrect

**Review by alessioarcara — APPROVED** [review]
Even though I believe there are some issues with this PR, I don't want to block this PR in order to not block other devs as well, with things that could be fixed with a further improvement branch.
Furthermore, the fact that you took 4 tasks and combined them into one has been a big mistake. It has complicated the review process and blocked other developers for this large pull request. I have also submitted large pull requests with many changed files, but the implemented task was one and contained enough to not block other developers.

File: src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java (hunk 4 above)
Comment (Matte0Sacc0) [reply src/main/java/com/aadm/cardexchange/server/services/ExchangeServiceImpl.java]: Seeing that it's easy to fix this issue and I have to resolve merge conflicts I'll fix it in the next push
